ee/apps/omnichannel-transcript/src/OmnichannelTranscript.ts

```diff
@@ -1,16 +1,272 @@
+		if (this.maxNumberOfConcurrentJobs <= this.currentJobNumber) {
+			this.log.error(`Processing transcript for room ${details.rid} by user ${details.userId} - Too many concurrent jobs, queuing again`);
+			throw new Error('retry');
+		}
+		this.currentJobNumber++;
+
+		const room = await LivechatRooms.findOneById(details.rid);
+		if (!room) {
+			throw new Error('room-not-found');
+		}
+		const messages = await this.getMessagesFromRoom({ rid: room._id });
+
+		const data = {
+			visitor:
+				room.v && (await LivechatVisitors.findOneById(room.v._id, { projection: { _id: 1, name: 1, username: 1, visitorEmails: 1 } })),
+			agent: room.servedBy && (await Users.findOneAgentById(room.servedBy._id, { projection: { _id: 1, name: 1, username: 1 } })),
+			closedAt: room.closedAt,
+			messages: await this.getFiles(details.userId, messages),
+		};
+
+		try {
+			await this.doRender({ template, data, details });
+		} catch (error) {
+			await this.pdfFailed({ details, e: error as Error });
+		}
+	}
+
+	async doRender({ template, data, details }: { template: Templates; data: WorkerData; details: WorkDetailsWithSource }): Promise<void> {
+		const buf: Uint8Array[] = [];
+		let outBuff = Buffer.alloc(0);
+
+		const stream = await this.worker.renderToStream({ template, data });
+		stream.on('data', (chunk) => {
+			buf.push(chunk);
+		});
+		stream.on('end', () => {
+			outBuff = Buffer.concat(buf);
+
+			return this.uploadService
+				.uploadFile({
+					userId: details.userId,
+					buffer: outBuff,
+					details: {
+						name: 'transcript.pdf',
+						type: 'application/pdf',
+						rid: details.rid,
+						// Rocket.cat is the goat
+						userId: 'rocket.cat',
+						size: outBuff.length,
+					},
+				})
+				.then((file) => this.pdfComplete({ details, file }))
+				.catch((e) => this.pdfFailed({ details, e }));
+		});
+	}
+
+	private async pdfFailed({ details, e }: { details: WorkDetailsWithSource; e: Error }): Promise<void> {
+		this.currentJobNumber--;
+
+		this.log.error(`Transcript for room ${details.rid} by user ${details.userId} - Failed: ${e.message}`);
+		const room = await LivechatRooms.findOneById(details.rid);
+		if (!room) {
+			return;
+		}
+		const user = await Users.findOneById(details.userId);
+		if (!user) {
+			return;
+		}
+
+		// Remove `transcriptRequestedPdf` from room to allow another request
+		await LivechatRooms.unsetTranscriptRequestedPdfById(details.rid);
+
+		const { rid } = await this.messageService.createDirectMessage({ to: details.userId, from: 'rocket.cat' });
+		this.log.log(`Transcript for room ${details.rid} by user ${details.userId} - Sending error message to user`);
+		await this.messageService.sendMessage({
+			fromId: 'rocket.cat',
+			rid,
+			msg: `${await this.translationService.translate('pdf_error_message', user)}: ${e.message}`,
+		});
+	}
+
+	private async pdfComplete({ details, file }: { details: WorkDetailsWithSource; file: IUpload }): Promise<void> {
+		this.currentJobNumber--;
```

Comment:
```suggestion
```

Won't be needed if you decide to go this route -> https://github.com/RocketChat/Rocket.Chat/pull/27568/files#r1057216219